### PR TITLE
tests for projects endpoint

### PR DESCRIPTION
**Pull request by Vital7788**

### CODE REVIEW

File: backend/app/tests/test_routers/test_projects.py [hunk 1]
```diff
@@ -1,60 +1,269 @@
...
+        await update(edition, self.session)
+
+        skill_generator = SkillGenerator(self.session)
+        skills = skill_generator.generate_skills()
+        skill_generator.add_to_db()
+        await self.session.commit()
+
+        project_skills = [{"skill_name": skill.name, "number": 3} for skill in skills]
+
+        project_coach = await read_where(User, User.role == UserRole.COACH, session=self.session)
```
Comment (StefVdHaute) [inline backend/app/tests/test_routers/test_projects.py:76]: There are multiple coaches in the testdata. Use get_user_by_name("user_approved_coach") or make read_where more detailed by checking whether user is active, approved and not disabled.

File: backend/app/tests/test_routers/test_projects.py [hunk 2]
```diff
@@ -1,60 +1,269 @@
...
+        project_id = ""
+        for user_title in allowed_users:
+            project_id = json.loads(responses1.get(user_title).content)["data"]["id"].split("/")[-1]
+
+        # Test whether created project is in the database
+        project_in_db = await read_where(Project, Project.name == project_name, session=self.session)
+        self.assertIsNotNone(project_in_db, f"'{project_name}' was not found in the database.")
+
+        # Project id in the database should be same as it in the response url
+        self.assertEqual(str(project_in_db.id), project_id, f"Project id of '{project_name}' in the response url is not same as it in the database.")
```
Comment (StefVdHaute) [inline backend/app/tests/test_routers/test_projects.py:105]: You made a sentence structure mistake. Consider `f"Project id of '{project_name}' in the response URL is not equal to the id found in the database."`.

File: backend/app/tests/test_routers/test_projects.py [hunk 3]
```diff
@@ -1,60 +1,269 @@
...
+
+        # test project user
+        project_coaches_in_db = await read_all_where(ProjectCoach, ProjectCoach.project_id == int(project_id), session=self.session)
+        self.assertIsNotNone(project_coaches_in_db, f"User has not been added for project {project_name} ")
+
+        project_skills_in_db = await read_all_where(ProjectRequiredSkill, ProjectRequiredSkill.project_id == int(project_id), session=self.session)
+        self.assertGreater(len(project_coaches_in_db), 0, f"Skill has not been added for project {project_name}")
+
+        # test project required skills
+        self.assertEqual(len(project_skills_in_db), len(skills), f"Not all skills have not been added for project {project_name}")
```
Comment (StefVdHaute) [inline backend/app/tests/test_routers/test_projects.py:125]: Double negation in sentence.

**Review by StefVdHaute — CHANGES_REQUESTED** [review]
I noticed a lot of your code is duplicated. Some general tips i would like you to implement:
1. You often use the same arguments for mock projects. You can save the arguments in a dict in the class and reuse them by using `Project(**the_arg_dict)`, this dict can also be used for your request body's.
2. If you want to seperate tests of different users don't write the tests twice. Write the test for 1 user (ex. user_admin) and generalize it so you could pass the users with equal access rights you want to test for to the testcode.
3. Your checks are very similar to the point my IDE gives warnings for multiple 20+ line blocks of duplicated code. Create a check function.

Lastly check my previous request for change.

**Comment by Vital7788** [comment]
requested changes have been made

**Comment by Vital7788** [comment]
> I noticed a lot of your code is duplicated. Some general tips i would like you to implement:
> 
>     1. You often use the same arguments for mock projects. You can save the arguments in a dict in the class and reuse them by using `Project(**the_arg_dict)`, this dict can also be used for your request body's.
> 
>     2. If you want to seperate tests of different users don't write the tests twice. Write the test for 1 user (ex. user_admin) and generalize it so you could pass the users with equal access rights you want to test for to the testcode.
> 
>     3. Your checks are very similar to the point my IDE gives warnings for multiple 20+ line blocks of duplicated code. Create a check function.
> 
> 
> Lastly check my previous request for change.

For your second point, I think having a single test to test all possible users makes the tests much harder to read and understand. Additionally, it's often not possible to generalize tests like that since multiple coaches could have access, but have different responses depending on which editions/projects they have access to.

All the other changes have been implemented.

**Review by Vital7788 — COMMENTED** [review]
looks good to me

**Pull request merged**